File: app/src/main/java/com/babestudios/companyinfouk/ui/officerdetails/OfficerDetailsActivity.java
package com.babestudios.companyinfouk.ui.officerdetails;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.babestudios.companyinfouk.CompaniesHouseApplication;
import com.babestudios.companyinfouk.R;
import com.babestudios.companyinfouk.data.model.officers.OfficerItem;
import com.babestudios.companyinfouk.ui.officerappointments.OfficerAppointmentsActivity;
import com.babestudios.companyinfouk.uiplugins.BaseActivityPlugin;
import com.google.gson.Gson;
import com.pascalwelsch.compositeandroid.activity.CompositeActivity;

import net.grandcentrix.thirtyinch.plugin.TiActivityPlugin;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class OfficerDetailsActivity extends CompositeActivity implements OfficerDetailsActivityView {

	@BindView(R.id.toolbar)
	Toolbar toolbar;

	@BindView(R.id.progressbar)
	ProgressBar progressbar;

	@BindView(R.id.textViewName)
	TextView textViewName;
	@BindView(R.id.textViewAppointedOn)
	TextView textViewAppointedOn;
	@BindView(R.id.textViewNationality)
	TextView textViewNationality;
	@BindView(R.id.textViewOccupation)
	TextView textViewOccupation;
	@BindView(R.id.textViewDateOfBirth)
	TextView textViewDateOfBirth;
	@BindView(R.id.textViewCountryOfResidence)
	TextView textViewCountryOfResidence;
	@BindView(R.id.textViewAddressLine1)
	TextView textViewAddressLine1;
	@BindView(R.id.textViewLocality)
	TextView textViewLocality;
	@BindView(R.id.textViewPostalCode)
	TextView textViewPostalCode;
	@BindView(R.id.textViewRegion)
	TextView textViewRegion;
	@BindView(R.id.textViewCountry)
	TextView textViewCountry;

	String officerItemString;
	String officerId;

	@Inject
	public OfficerDetailsPresenter officerDetailsPresenter;

	TiActivityPlugin<OfficerDetailsPresenter, OfficerDetailsActivityView> officerDetailsActivityPlugin = new TiActivityPlugin<>(
			() -> {
				CompaniesHouseApplication.getInstance().getApplicationComponent().inject(this);
				return officerDetailsPresenter;
			});

	BaseActivityPlugin baseActivityPlugin = new BaseActivityPlugin();

	public OfficerDetailsActivity() {
		addPlugin(officerDetailsActivityPlugin);
		addPlugin(baseActivityPlugin);
	}

	@SuppressWarnings("ConstantConditions")
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_officer_details);
		baseActivityPlugin.logScreenView(this.getLocalClassName());

		ButterKnife.bind(this);
		officerItemString = getIntent().getStringExtra("officerItem");
		Gson gson = new Gson();
		OfficerItem officerItem = gson.fromJson(officerItemString, OfficerItem.class);
		textViewName.setText(officerItem.name);
		textViewAppointedOn.setText(officerItem.appointedOn);
		textViewNationality.setText(officerItem.nationality);
		textViewOccupation.setText(officerItem.occupation);
		if(officerItem.dateOfBirth != null) {
			textViewDateOfBirth.setText(officerItem.dateOfBirth.month + "/" + officerItem.dateOfBirth.year);
		} else {
			textViewDateOfBirth.setText(R.string.officer_details_unknown);
		}
		textViewCountryOfResidence.setText(officerItem.countryOfResidence);
		textViewAddressLine1.setText(officerItem.address.addressLine1);
		textViewLocality.setText(officerItem.address.locality);
		textViewPostalCode.setText(officerItem.address.postalCode);
		if(officerItem.address.region == null) {
			textViewRegion.setVisibility(View.GONE);
		} else {
			textViewRegion.setVisibility(View.VISIBLE);
			textViewRegion.setText(officerItem.address.region);
		}
		if(officerItem.address.country == null) {
			textViewCountry.setVisibility(View.GONE);
		} else {
			textViewCountry.setVisibility(View.VISIBLE);
			textViewCountry.setText(officerItem.address.country);
		}

		if (toolbar != null) {
			setSupportActionBar(toolbar);
			getSupportActionBar().setDisplayHomeAsUpEnabled(true);
			getSupportActionBar().setTitle(R.string.officer_details);
			toolbar.setNavigationOnClickListener(v -> onBackPressed());
		}
		Pattern pattern = Pattern.compile("officers/(.+)/appointments");
		Matcher matcher = pattern.matcher(officerItem.links.officer.appointments);
		if (matcher.find())
		{
			officerId = matcher.group(1);
		}
	}

	@OnClick(R.id.buttonAppointments)
	public void onClick() {
		Intent intent = new Intent(this, OfficerAppointmentsActivity.class);
		intent.putExtra("officerId", officerId);
		baseActivityPlugin.startActivityWithRightSlide(intent);
	}

	@Override
	public void showProgress() {
		progressbar.setVisibility(View.VISIBLE);
	}

	@Override
	public void hideProgress() {
		progressbar.setVisibility(View.GONE);
	}

	@Override
	public void super_onBackPressed() {
		super.super_finish();
		super_overridePendingTransition(R.anim.left_slide_in, R.anim.left_slide_out);
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/officerdetails/OfficerDetailsActivity.kt
package com.babestudios.companyinfouk.ui.officerdetails

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View

import kotlinx.android.synthetic.main.activity_officer_details.*
import com.babestudios.companyinfouk.R
import com.uber.autodispose.AutoDispose
import com.uber.autodispose.ScopeProvider
import com.ubercab.autodispose.rxlifecycle.RxLifecycleInterop
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity
import io.reactivex.CompletableSource
import androidx.lifecycle.ViewModelProviders
import com.babestudios.companyinfouk.Injector
import com.babestudios.companyinfouk.data.model.officers.OfficerItem
import com.babestudios.companyinfouk.ext.startActivityWithRightSlide
import com.babestudios.companyinfouk.ui.officerappointments.createOfficerAppointmentsIntent
import com.jakewharton.rxbinding2.view.RxView

import io.reactivex.disposables.CompositeDisposable

private const val OFFICER_ITEM = "com.babestudios.companyinfouk.ui.officer_item"

class OfficerDetailsActivity : RxAppCompatActivity(), ScopeProvider {


	override fun requestScope(): CompletableSource = RxLifecycleInterop.from(this).requestScope()

	private val viewModel by lazy { ViewModelProviders.of(this).get(OfficerDetailsViewModel::class.java) }

	private lateinit var officerDetailsPresenter: OfficerDetailsPresenterContract

	private val eventDisposables: CompositeDisposable = CompositeDisposable()

	//region life cycle

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_officer_details)
		setSupportActionBar(pabOfficerDetails.getToolbar())
		supportActionBar?.setDisplayHomeAsUpEnabled(true)
		pabOfficerDetails.setNavigationOnClickListener { onBackPressed() }
		supportActionBar?.setTitle(R.string.officer_details)
		when {
			viewModel.state.value.officerItem != null -> {
				initPresenter(viewModel)
			}
			savedInstanceState != null -> {
				savedInstanceState.getParcelable<OfficerDetailsState>("STATE")?.let {
					with(viewModel.state.value) {

						officerItem = it.officerItem
					}
				}
				initPresenter(viewModel)
			}
			else -> {
				viewModel.state.value.officerItem = intent.getParcelableExtra(OFFICER_ITEM)
				initPresenter(viewModel)
			}
		}

		observeState()
	}

	override fun onResume() {
		super.onResume()
		observeActions()
	}

	override fun onSaveInstanceState(outState: Bundle) {
		outState.putParcelable("STATE", viewModel.state.value)
		super.onSaveInstanceState(outState)
	}

	private fun initPresenter(viewModel: OfficerDetailsViewModel) {
		if (!::officerDetailsPresenter.isInitialized) {
			officerDetailsPresenter = Injector.get().officerDetailsPresenter()
			officerDetailsPresenter.setViewModel(viewModel, requestScope())
		}
	}

	override fun onBackPressed() {
		super.onBackPressed()
		overridePendingTransition(R.anim.left_slide_in, R.anim.left_slide_out)
	}

	//endregion

	//region render

	private fun observeState() {
		viewModel.state
				.`as`(AutoDispose.autoDisposable(this))
				.subscribe { render(it) }
	}

	private fun render(state: OfficerDetailsState) {
		when {
			state.contentChange == ContentChange.OFFICER_ITEM_RECEIVED -> {
				state.contentChange = ContentChange.NONE
				showOfficerDetails()
			}
		}
	}

	private fun showOfficerDetails() {
		val officerItem = viewModel.state.value.officerItem
		textViewName?.text = officerItem?.name
		textViewAppointedOn?.text = officerItem?.appointedOn
		textViewNationality?.text = officerItem?.nationality
		textViewOccupation?.text = officerItem?.occupation
		officerItem?.dateOfBirth?.let {
			textViewDateOfBirth?.text = "${it.month.toString()} / ${it.year.toString()}"
		} ?: run {
			textViewDateOfBirth?.setText(R.string.officer_details_unknown)
		}
		textViewCountryOfResidence?.text = officerItem?.countryOfResidence
		textViewAddressLine1?.text = officerItem?.address?.addressLine1
		textViewLocality?.text = officerItem?.address?.locality
		textViewPostalCode?.text = officerItem?.address?.postalCode
		officerItem?.address?.region?.let {
			textViewRegion?.visibility = View.VISIBLE
			textViewRegion?.text = it
		} ?: run {
			textViewRegion?.visibility = View.GONE
		}
		officerItem?.address?.country?.let {
			textViewCountry?.text = it
			textViewCountry?.visibility = View.VISIBLE
		} ?: run {
			textViewCountry?.visibility = View.GONE
		}
	}

	//endregion

	//region events

	private fun observeActions() {
		eventDisposables.clear()
		RxView.clicks(buttonAppointments)
				.`as`(AutoDispose.autoDisposable(this))
				.subscribe { startActivityWithRightSlide(createOfficerAppointmentsIntent(viewModel.state.value.officerId)) }

	}

	//endregion
}

fun Context.createOfficerDetailsIntent(officerItem: OfficerItem): Intent {
	return Intent(this, OfficerDetailsActivity::class.java)
			.putExtra(OFFICER_ITEM, officerItem)
}
